### PR TITLE
[WIP] Parallelize DecisionTree training.

### PR DESCRIPTION
**Pull request by RishabhGarg108**
This PR attempts to make the training of DecisionTree parallel.

CC: @rcurtin @gmanlan

### CODE REVIEW

File: src/mlpack/methods/decision_tree/all_dimension_select.hpp [hunk 1]
```diff
@@ -25,35 +25,24 @@ class AllDimensionSelect
   /**
    * Construct the AllDimensionSelect object for the given number of dimensions.
    */
-  AllDimensionSelect() : i(0), dimensions(0) { }
+  AllDimensionSelect() : dimensions(0) { }
 
   /**
-   * Get the first dimension to select from.
+   * Get the current dimension.
```
Comment (rcurtin) [inline src/mlpack/methods/decision_tree/all_dimension_select.hpp:31]:
```suggestion
   * Get the `i`th dimension.
```
Just a little pedantic comment fix. :)

**Review by rcurtin — COMMENTED** [review]
Looks great!  Now I think you can just add the OpenMP arg-max + max reduction and it'll be good to go.  (I dunno if "arg-max + max" makes sense, but I guess, we have to keep two variables in the reduction---the max (`bestGain`) and the arg-max (`bestDim`).)

File: src/mlpack/methods/decision_tree/multiple_random_dimension_select.hpp [hunk 2]
```diff
@@ -30,20 +30,25 @@ class MultipleRandomDimensionSelect
...
   /**
-   * Get the first random value.
+   * Get the current dimension.
    */
-  size_t Begin()
+  size_t GetDimension(size_t i) const { return values[i]; }
+
+  /**
+   * Populates the set of randomly selected dimensions and 
+   * return the total number of dimensions to iterate.
```
Comment (zoq) [inline src/mlpack/methods/decision_tree/multiple_random_dimension_select.hpp:43]:
```suggestion
   * returns the total number of dimensions to iterate.
```

File: src/mlpack/methods/decision_tree/decision_tree_regressor_impl.hpp [hunk 3]
```diff
@@ -811,36 +811,49 @@ double DecisionTreeRegressor<FitnessFunction,
...
-      // If the splitter did not report that it improved, then move to the next
-      // dimension.
-      if (dimGain == DBL_MAX)
-        continue;
-
-      bestDim = i;
-      bestGain = dimGain;
+      double localBestGain = 0;
+      size_t localBestDim = 0;
+      #pragma omp for reduction( max : bestGain)
```
Comment (zoq) [inline src/mlpack/methods/decision_tree/decision_tree_regressor_impl.hpp:822]: Maybe I missed it, but did you already run some benchmark against the parallelization code?

**Pull request closed without merging**